sound/soc/intel/boards/sof_sdw.c

```diff
@@ -1141,9 +1141,10 @@ static int sof_card_dai_links_create(struct snd_soc_card *card)
 	}
 	/*
 	 * mach_params->dmic_num will be used to set the cfg-mics value of card->components
-	 * string. Overwrite it to the actual number of PCH DMICs used in the device.
+	 * string. And dmic_num is the number of the dmic dai links. Overwrite it if dmic_num
+	 * is set because of the SOC_SDW_PCH_DMIC quirk.
 	 */
-	mach_params->dmic_num = dmic_num;
+	mach_params->dmic_num = mach_params->dmic_num ? mach_params->dmic_num : dmic_num;
```

Comment:
I think this is still a little broken since it will leave mach_params->dmic_num set, thus adding "mic:dmic cfg-mic:" even in the case ignore_intel_dmic is set, how about this as an alternative:
```
        /* enable dmic01 & dmic16k */
        if (ctx->ignore_internal_dmic) {
                dev_warn(dev, "Ignoring PCH DMIC\n");
                mach_params->dmic_num = 0;
        } else if (sof_sdw_quirk & SOC_SDW_PCH_DMIC) {
                /*
                 * mach_params->dmic_num will be used to set the cfg-mics value
                 * of card->components string. Overwrite it to the actual number
                 * of PCH DMICs used in the device.
                 */
                dmic_num = 2;
                mach_params->dmic_num = dmic_num;
        } else if (mach_params->dmic_num) {
                dmic_num = 2;
        }
```
Not totally sure whether the quirk or mach_params->dmic_num should take precedence, but I went with the quirk since I assume it should not be present unless one wants to use it.